source/opt/decompose_initialized_variables_pass.cpp

```diff
@@ -57,20 +72,41 @@ Pass::Status DecomposeInitializedVariablesPass::Process() {
       Instruction* store_inst = new Instruction(
           context(), SpvOpStore, 0, 0,
           {{SPV_OPERAND_TYPE_ID, {var_id}}, {SPV_OPERAND_TYPE_ID, {val_id}}});
-      new_stores.push_back(store_inst);
+      function_stores.push_back(store_inst);
       iter->RemoveOperand(3);
-      get_def_use_mgr()->UpdateDefUse(&*iter);
     }
 
-    for (auto store = new_stores.begin(); store != new_stores.end(); ++store) {
+    if (entry_ids.find(func->result_id()) != entry_ids.end()) {
+      for (auto store_ids : global_stores) {
+        uint32_t var_id;
+        uint32_t val_id;
+        std::tie(var_id, val_id) = store_ids;
+        auto* store_inst = new Instruction(
+            context(), SpvOpStore, 0, 0,
+            {{SPV_OPERAND_TYPE_ID, {var_id}}, {SPV_OPERAND_TYPE_ID, {val_id}}});
+        context()->AnalyzeDefUse(store_inst);
+        context()->set_instr_block(store_inst, &*first_block);
+        first_block->AddInstruction(std::unique_ptr<Instruction>(store_inst));
+        if (last_var == first_block->end()) {
```

Comment:
You can probably simplify a lot of the code related to `last_var` if you changed it from pointing to the last variable to point to the first instruction after the last variable.  That instruction will always exist, and then there is not special case for when there are no variable.  You can just always insert before.